local/network.go

```diff
@@ -737,3 +592,219 @@ func createFileAndWrite(path string, contents []byte) error {
+	port, err = getFreePort()
+	if err != nil {
+		return 0, fmt.Errorf("couldn't get free API port: %w", err)
+	}
+	return port, nil
+}
+
+// nodeInfo is only internally used so we can pass some node information between funcs
+type nodeInfo struct {
+	flags   []string
+	apiPort uint16
+	p2pPort uint16
+	nodeID  ids.ShortID
+}
+
+// buildFlags returns the:
+// 1) Flags
+// 2) API port
+// 3) P2P port
+// 4) Node ID
+// of the node being added with config [nodeConfig], config file [configFile],
+// and directory at [nodeDir].
+func (ln *localNetwork) buildFlags(configFile map[string]interface{}, nodeDir string, nodeConfig *node.Config) (nodeInfo, error) {
+	// Add flags in [ln.Flags] to [nodeConfig.Flags]
+	// Assumes [nodeConfig.Flags] is non-nil
+	addNetworkFlags(ln.log, ln.flags, nodeConfig.Flags)
+
+	// Tell the node to put the database in [nodeDir] unless given in config file
+	dbPath, err := getConfigEntry(configFile, config.DBPathKey, nodeDir)
+	if err != nil {
+		return nodeInfo{}, err
+	}
+
+	// Tell the node to put the log directory in [nodeDir/logs] unless given in config file
+	logsDir, err := getConfigEntry(configFile, config.LogsDirKey, filepath.Join(nodeDir, "logs"))
+	if err != nil {
+		return nodeInfo{}, err
+	}
+
+	// Use random free API port unless given in config file
+	apiPort, err := getPort(configFile, config.HTTPPortKey)
+	if err != nil {
+		return nodeInfo{}, err
+	}
+
+	// Use a random free P2P (staking) port unless given in config file
+	// Use random free API port unless given in config file
+	p2pPort, err := getPort(configFile, config.StakingPortKey)
+	if err != nil {
+		return nodeInfo{}, err
+	}
+
+	// Flags for AvalancheGo
+	flags := []string{
+		fmt.Sprintf("--%s=%d", config.NetworkNameKey, ln.networkID),
+		fmt.Sprintf("--%s=%s", config.DBPathKey, dbPath),
+		fmt.Sprintf("--%s=%s", config.LogsDirKey, logsDir),
+		fmt.Sprintf("--%s=%d", config.HTTPPortKey, apiPort),
+		fmt.Sprintf("--%s=%d", config.StakingPortKey, p2pPort),
+		fmt.Sprintf("--%s=%s", config.BootstrapIPsKey, ln.bootstrapIPs),
+		fmt.Sprintf("--%s=%s", config.BootstrapIDsKey, ln.bootstrapIDs),
+	}
+	// Write staking key/cert etc. to disk so the new node can use them,
+	// and get flag that point the node to those files
+	fileFlags, err := writeFiles(ln.genesis, nodeDir, nodeConfig)
+	if err != nil {
+		return nodeInfo{}, err
+	}
+	flags = append(flags, fileFlags...)
+
+	// Add flags given in node config.
+	// Note these will overwrite existing flags if the same flag is given twice.
+	for flagName, flagVal := range nodeConfig.Flags {
+		if _, ok := warnFlags[flagName]; ok {
+			ln.log.Warn("The flag %s has been provided. This can create conflicts with the runner. The suggestion is to remove this flag", flagName)
+		}
+		flags = append(flags, fmt.Sprintf("--%s=%v", flagName, flagVal))
+	}
+
+	// Parse this node's ID
+	nodeID, err := utils.ToNodeID([]byte(nodeConfig.StakingKey), []byte(nodeConfig.StakingCert))
+	if err != nil {
```

Comment:
I should have leave this outside of this function. Probably a new function that generates the ports outside of this, can also return a partial localNode (ports, nodeId). initLocalNode?